src/Sylius/Bundle/ApiBundle/spec/Validator/Constraints/ChosenPaymentRequestActionEligibilityValidatorSpec.php

```diff
@@ -0,0 +1,270 @@
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace spec\Sylius\Bundle\ApiBundle\Validator\Constraints;
+
+use PhpSpec\ObjectBehavior;
+use Sylius\Bundle\ApiBundle\Command\Payment\AddPaymentRequest;
+use Sylius\Bundle\ApiBundle\Validator\Constraints\ChosenPaymentRequestActionEligibility;
+use Sylius\Bundle\PaymentBundle\CommandProvider\PaymentRequestCommandProviderInterface;
+use Sylius\Bundle\PaymentBundle\CommandProvider\ServiceProviderAwareCommandProviderInterface;
+use Sylius\Bundle\PaymentBundle\Provider\GatewayFactoryNameProviderInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+use Sylius\Component\Core\Repository\PaymentMethodRepositoryInterface;
+use Sylius\Component\Payment\Model\GatewayConfigInterface;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidatorInterface;
+use Symfony\Component\Validator\Context\ExecutionContextInterface;
+
+final class ChosenPaymentRequestActionEligibilityValidatorSpec extends ObjectBehavior
+{
+    function let(
+        PaymentMethodRepositoryInterface $paymentMethodRepository,
+        ServiceProviderAwareCommandProviderInterface $gatewayFactoryCommandProvider,
+        GatewayFactoryNameProviderInterface $gatewayFactoryNameProvider,
+        ExecutionContextInterface $executionContext,
+    ): void {
+        $this->beConstructedWith($paymentMethodRepository, $gatewayFactoryCommandProvider, $gatewayFactoryNameProvider);
+
+        $this->initialize($executionContext);
+    }
+
+    function it_is_a_constraint_validator(): void
+    {
+        $this->shouldImplement(ConstraintValidatorInterface::class);
+    }
+
+    function it_throws_an_exception_if_value_is_not_a_add_payment_request(): void
+    {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', ['', new ChosenPaymentRequestActionEligibility()])
+        ;
+    }
+
+    function it_throws_an_exception_if_constraint_is_not_an_instance_of_chosen_payment_request_action_eligibility(): void
+    {
+        $this
+            ->shouldThrow(\InvalidArgumentException::class)
+            ->during('validate', [
+                new AddPaymentRequest('ORDER_TOKEN', 123, 'PAYMENT_METHOD_CODE'),
+                new class() extends Constraint {
+                },
+            ])
+        ;
+    }
+
+    function it_does_nothing_if_there_no_action_given(
+        ExecutionContextInterface $executionContext,
+    ): void {
+        $command = new AddPaymentRequest(
+            orderTokenValue: 'ORDER_TOKEN',
+            paymentId: 123,
+            paymentMethodCode: 'PAYMENT_METHOD_CODE',
+        );
+
+        $executionContext
+            ->addViolation('sylius.payment_method.not_exist', ['%code%' => 'PAYMENT_METHOD_CODE'])
+            ->shouldNotBeCalled()
+        ;
```

Comment:
Maybe instead of this one, we should check other service
```suggestion
        $paymentMethodRepository->findOneBy(Argument::any())->shouldNotBeCalled();
```
🤔